snuba/query/__init__.py

```diff
@@ -44,16 +45,17 @@ class OrderBy:
 @dataclass(frozen=True)
 class SelectedExpression:
     # The name of this column in the resultset.
+    # TODO: Make this non nullable
     name: Optional[str]
     expression: Expression
 
 
 TExp = TypeVar("TExp", bound=Expression)
 
 
-class Query(ABC, Generic[TDataSource]):
+class Query(DataSource, Generic[TDataSource]):
```

Comment:
I tried a few times to keep this abstract and to have two concrete subclasses (Composite and Simple) which would be a better design because it would force a simple query to actually be a ProcessableQuery, but I did not manage to keep the composite query parametric to the type of data source of the simple query contained inside (nested potentially multiple level) and still allowing it to be a data source of another query.

Essentially the type parameter would have been something like
```
TDataSource = TypeVar("TDataSource", bound=Union[SimpleQuery, CompositeQuery[TDataSourcve]]  
# which is invalid.

# or
TDataSource = TypeVar("TDataSource", bound=Union[SimpleQuery, Query[TSomeCompositeSource]]  
# which is invalid as well because TSomeCompositeSource is still defined in terms of the Query class.
```

I think I will eventually try again to make this class abstract again. IT does not seem a big deal since queries are produced in one place only (the parser).

Reply:
Ignore that, I found a way.